File: RudderUnityPluginDemo/Assets/Plugins/iOS/RudderSDKUnity/RudderClient.m
//
//  RudderClient.m
//  RudderSDKUnity
//
//  Created by Arnab Pal on 22/10/19.
//  Copyright © 2019 Rudderlabs. All rights reserved.
//

#import "RudderClient.h"
#import "EventRepository.h"
#import "RudderConfig.h"
#import "RudderMessage.h"
#import "RudderMessageBuilder.h"

static RudderClient *_instance = nil;
static EventRepository *_repository = nil;

@implementation RudderClient

+ (void)_initiateInstance:(NSString *)writeKey
              endPointUrl:(NSString *)endPointUrl
           flushQueueSize:(int)flushQueueSize
         dbCountThreshold:(int)dbCountThreshold
             sleepTimeout:(int)sleepTimeout
                 logLevel:(int)logLevel {
    if (_instance == nil) {
        static dispatch_once_t onceToken;
        dispatch_once(&onceToken, ^{
            _instance = [[self alloc] init];
            
            // create RudderConfig based on the provided parameters
            RudderConfig *config = [[RudderConfig alloc] init];
            config.endPointUrl = endPointUrl;
            config.flushQueueSize = flushQueueSize;
            config.dbCountThreshold = dbCountThreshold;
            config.sleepTimeout = sleepTimeout;
            config.logLevel = logLevel;
            
            // initiate repository with the config
            _repository = [EventRepository initiate:writeKey config:config];
        });
    }
}

+ (void) _logEvent:(NSString *)eventType
         eventName:(NSString *)eventName
            userId:(NSString *)userId
eventPropertiesJson:(NSString *)eventPropertiesJson
userPropertiesJson:(NSString *)userPropertiesJson
  integrationsJson:(NSString *)integrationsJson {
    // create error instance
    NSError *error;
    
    // create messageBuilder
    RudderMessageBuilder *builder = [[RudderMessageBuilder alloc] init];
    [builder setEventName:eventName];
    
    // create eventProperties dict from json and set it to builder
    NSDictionary *eventProperties = [NSJSONSerialization JSONObjectWithData:[eventPropertiesJson dataUsingEncoding:NSUTF8StringEncoding] options:NSJSONReadingMutableContainers error:&error];
    [builder setPropertyDict:eventProperties];
    
    // create userProperties dict from json and set it to builder
    NSDictionary *userProperties = [NSJSONSerialization JSONObjectWithData:[userPropertiesJson dataUsingEncoding:NSUTF8StringEncoding] options:NSJSONReadingMutableContainers error:&error];;
    [builder setUserProperty:userProperties];

    // set userId to the builder
    [builder setUserId:userId];
    
    // finally build the message
    RudderMessage *message = [builder build];
    NSDictionary *integrations = [NSJSONSerialization JSONObjectWithData:[integrationsJson dataUsingEncoding:NSUTF8StringEncoding] options:NSJSONReadingMutableContainers error:&error];;
    message.integrations = integrations;
    
    // set the message type after building the message
    message.type = eventType;
    
    // dump the message if repository is initiated
    if (_repository != nil) {
        [_repository dump:message];
    }
}

@end

File: Plugins/iOS/RudderPlugin_iOS/RudderPlugin_iOS/Source/RudderClient.swift
//
//  RudderClient.swift
//  RudderPlugin_iOS
//
//  Created by Arnab Pal on 14/09/19.
//  Copyright © 2019 Arnab Pal. All rights reserved.
//

import Foundation

class RudderClient: NSObject {
    private static var instance: RudderClient? = nil
    private var repository: EventRepository? = nil
    
    private override init() {
        // do nothing
    }
    
    init(_writeKey: String, _endPointUrl: String, _flushQueueSize: Int32, _dbCountThreshold: Int32, _sleepTimeOut: Int32) {
        self.repository = EventRepository.getInstance(writeKey: _writeKey, endPointUrl: _endPointUrl, flushQueueSize: _flushQueueSize, dbCountThreshold: _dbCountThreshold, sleepTimeOut: _sleepTimeOut)
    }
    
    @objc static func getInstance() -> RudderClient? {
        return instance
    }
    
    @objc static func getInstance(writeKey: String) -> RudderClient {
        return getInstance(writeKey: writeKey, endPointUrl: Constants.BASE_URL)
    }
    
    @objc static func getInstance(writeKey: String, endPointUrl: String) -> RudderClient {
        return getInstance(writeKey: writeKey, endPointUrl: endPointUrl, flushQueueSize: Constants.FLUSH_QUEUE_SIZE)
    }
    
    @objc static func getInstance(writeKey: String, endPointUrl: String, flushQueueSize: Int32) -> RudderClient {
        return getInstance(writeKey: writeKey, endPointUrl: endPointUrl, flushQueueSize: flushQueueSize, dbCountThreshold: Constants.DB_COUNT_THRESHOLD, sleepTimeOut: Constants.SLEEP_TIME_OUT)
    }
    
    @objc static func getInstance(writeKey: String, endPointUrl: String, flushQueueSize: Int32, dbCountThreshold: Int32, sleepTimeOut: Int32) -> RudderClient {
        if (instance == nil) {
            instance = RudderClient(_writeKey: writeKey, _endPointUrl: endPointUrl, _flushQueueSize: flushQueueSize, _dbCountThreshold: dbCountThreshold, _sleepTimeOut: sleepTimeOut)
        }
        return instance!
    }
    
    @objc static func initiateInstance(writeKey: String, endPointUrl: String, flushQueueSize: Int32, dbCountThreshold: Int32, sleepTimeOut: Int32) {
        if (instance == nil) {
            instance = RudderClient(_writeKey: writeKey, _endPointUrl: endPointUrl, _flushQueueSize: flushQueueSize, _dbCountThreshold: dbCountThreshold, _sleepTimeOut: sleepTimeOut)
        }
    }
    
    @objc func track(element: RudderElement) {
        do {
            element.setEventType(type: MessageType.TRACK)
            try repository!.dump(element: element)
        } catch {
            print(error.localizedDescription)
        }
    }
    
    @objc func track(builder: RudderElementBuilder) {
        track(element: builder.build())
    }
    
    @objc func screen(element: RudderElement) {
        do {
            element.setEventType(type: MessageType.SCREEN)
            try repository!.dump(element: element)
        } catch {
            print(error.localizedDescription)
        }
    }
    
    @objc func screen(builder: RudderElementBuilder) {
        screen(element: builder.build())
    }
    
    @objc func page(element: RudderElement) {
        do {
            element.setEventType(type: MessageType.PAGE)
            try repository!.dump(element: element)
        } catch {
            print(error.localizedDescription)
        }
    }
    
    @objc func page(builder: RudderElementBuilder) {
        page(element: builder.build())
    }
    
    @objc func identify(element: RudderElement) {
        do {
            element.setEventType(type: MessageType.IDENTIFY)
            try repository!.dump(element: element)
        } catch {
            print(error.localizedDescription)
        }
    }
    
    @objc func identify(builder: RudderElementBuilder) {
        identify(element: builder.build())
    }
    
    @objc func logEvent(eventType: String, eventName: String, userId: String, eventPropertiesJson: String, userPropertiesJson: String, integrationsJson: String) {
        do {
            let element = RudderElementBuilder()
                .withEventName(eventName: eventName)
                .withUserId(userId: userId)
                .build()
            element.setEventType(type: eventType)
            element.setEventPropertiesString(eventProperties: eventPropertiesJson)
            element.setUserPropertiesString(userProperties: userPropertiesJson)
            element.setIntegrationsString(integrations: integrationsJson)
            try repository?.dump(element: element)
        } catch {
            print(error.localizedDescription)
        }
    }
    
    @objc func logEvent(eventType: String, eventName: String, userId: String, eventProperties: Dictionary<String, NSObject>, userProperties: Dictionary<String, NSObject>, integrations: Dictionary<String, Bool>) {
        do {
            let element = RudderElementBuilder()
                .withEventName(eventName: eventName)
                .withUserId(userId: userId)
                .withEventProperties(properties: eventProperties)
                .withUserProperties(properties: userProperties)
                .withIntegration(integrations: integrations)
                .build()
            element.setEventType(type: eventType)
            try repository?.dump(element: element)
        } catch {
            print(error.localizedDescription)
        }
    }
}
